pkg/inject/inject.go

```diff
@@ -0,0 +1,595 @@
+			patch.addVolumeRoot()
+		}
+		patch.addVolume(configMapVolume)
+		patch.addVolume(secretVolume)
+	}
+
+	patch.addContainer(&sidecar)
+
+	if !conf.globalConfig.GetCniEnabled() {
+		nonRoot := false
+		runAsUser := int64(0)
+		initContainer := &v1.Container{
+			Name:                     k8s.InitContainerName,
+			Image:                    conf.taggedProxyInitImage(),
+			ImagePullPolicy:          v1.PullPolicy(conf.proxyConfig.GetProxyInitImage().GetPullPolicy()),
+			TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+			Args:                     initArgs,
+			SecurityContext: &v1.SecurityContext{
+				Capabilities: &v1.Capabilities{
+					Add: []v1.Capability{v1.Capability("NET_ADMIN")},
+				},
+				Privileged:   &f,
+				RunAsNonRoot: &nonRoot,
+				RunAsUser:    &runAsUser,
+			},
+		}
+		if len(conf.podSpec.InitContainers) == 0 {
+			patch.addInitContainerRoot()
+		}
+		patch.addInitContainer(initContainer)
+	}
+}
+
+// Given a ObjectMeta, update ObjectMeta in place with the new labels and
+// annotations.
+func (conf *ResourceConfig) injectObjectMeta(patch *Patch) {
+	if len(conf.objMeta.Annotations) == 0 {
+		patch.addPodAnnotationsRoot()
+	}
+	patch.addPodAnnotation(k8s.CreatedByAnnotation, k8s.CreatedByAnnotationValue())
+	patch.addPodAnnotation(k8s.ProxyVersionAnnotation, conf.globalConfig.GetVersion())
+
+	if conf.globalConfig.GetIdentityContext() != nil {
+		patch.addPodAnnotation(k8s.IdentityModeAnnotation, k8s.IdentityModeOptional)
+	} else {
+		patch.addPodAnnotation(k8s.IdentityModeAnnotation, k8s.IdentityModeDisabled)
+	}
+
+	if len(conf.objMeta.Labels) == 0 {
+		patch.addPodLabelsRoot()
+	}
+	for k, v := range conf.podLabels {
+		patch.addPodLabel(k, v)
+	}
+}
+
+func (conf *ResourceConfig) taggedProxyImage() string {
+	return fmt.Sprintf("%s:%s",
+		conf.proxyConfig.GetProxyImage().GetImageName(),
+		conf.globalConfig.GetVersion())
+}
+
+func (conf *ResourceConfig) taggedProxyInitImage() string {
+	return fmt.Sprintf("%s:%s",
+		conf.proxyConfig.GetProxyInitImage().GetImageName(),
+		conf.globalConfig.GetVersion())
+}
+
+// shouldInject determines whether or not the given deployment should be
+// injected. A deployment shouldn't be injected if it contains any known sidecars
+// or is on a HostNetwork or the pod is not annotated with "linkerd.io/inject: disabled".
+// Additionally, a deployment should be injected if:
+// - old CLI inject logic: namespace annotations unavailable
+// - the deployment's namespace has the linkerd.io/inject annotation set to
+//   "enabled", and the deployment's pod spec does not have the
+//   linkerd.io/inject annotation set to "disabled"; or
+// - the deployment's pod spec has the linkerd.io/inject annotation set to "enabled"
+func (conf *ResourceConfig) shouldInject(r Report) bool {
+	if !r.Injectable() {
+		return false
+	}
+
+	podAnnotation := conf.objMeta.Annotations[k8s.ProxyInjectAnnotation]
+	if conf.nsAnnotations == nil {
+		return true
```

Comment:
should this check be removed? if `conf.nsAnnotations == nil` we shouldn't return true unless `podAnnotation == k8s.ProxyInjectEnabled`?

Reply:
Yeah, I think this must be the source of https://github.com/linkerd/linkerd2/pull/2334#discussion_r261472515, but I can't figure out what's up.

Only the auto-inject webhook calls `WithNsAnnotations`, so for the CLI inject code path, I'd expect `conf.nsAnnotations` to be `nil`. For CLI inject, we only care about pod annotations, so I'd expect this return statement to be:

```
		return podAnnotation != k8s.ProxyInjectDisabled
```

But it looks like it's also OK to `return true` as written, since the `Injectable` call above this is already checking that the `linkerd.io/inject: disabled` annotation is not set. This is pretty subtle though, so if there was some way to make this easier to reason about, that would be great.

For the auto-inject use case, we're expecting `conf.nsAnnotations` to not be `nil`, since we've called `WithNsAnnotations`, in which case we'll proceed to the namespace annotation check later in this method. But my guess is that's it's still possible for `conf.nsAnnotations` to be `nil` even after calling `WithNsAnnotations`, if the namespace doesn't actually have any annotations. If that were the case (untested), then that would explain why the proxy injector is injecting into namespaces that don't contain the `linkerd.io/inject: enabled` annotation.

Anyway, we might need to refactor this slightly to make it easier to understand how the inject decision is made.

Reply:
I've pushed #2433 as a suggestion to try making this more tractable. Although the functionality is the same, as I couldn't repro [#2334 (comment)](https://github.com/linkerd/linkerd2/pull/2334#discussion_r261472515).

Reply:
I also tried to repro this without success:
```bash
bin/linkerd install --proxy-auto-inject | kubectl apply -f -
curl https://run.linkerd.io/emojivoto.yml | kubectl apply -f -
```

Noting that emojivoto.yml does not have any `annotations` sections, it appears `conf.nsAnnotations` (coming from `namespace.GetAnnotations()`) is an empty `map[string]string` and not `nil`.

All that said, I'm onboard with making this code a bit more explicit, as I wouldn't want to rely on this subtlety when deciding whether to auto inject something. Also onboard with hashing this out in #2433 (thanks for posting that @alpeb).

Reply:
Hmm, ok, maybe it's something about my environment, but I can't repro the same behavior with `edge-19.2.5` -- it only happens on this branch.

In my environment, I have 2 namespaces configured, `kltest1` and `kltest2`, as follows:

```
$ kubectl get ns/kltest1 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    linkerd.io/inject: enabled
  creationTimestamp: 2019-02-12T17:18:23Z
  name: kltest1
  resourceVersion: "110658"
  selfLink: /api/v1/namespaces/kltest1
  uid: 3382bbc5-2eea-11e9-a483-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

$ kubectl get ns/kltest2 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: 2019-02-12T17:18:24Z
  name: kltest2
  resourceVersion: "110623"
  selfLink: /api/v1/namespaces/kltest2
  uid: 3447222e-2eea-11e9-a483-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

Only `kltest1` has the `linkerd.io/inject: enabled` annotation.

I then install `edge-19.2.5` with:

```
linkerd install --proxy-auto-inject --tls optional | kubectl apply -f -
```

And then apply the booksapp config in both namespaces:

```
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest1
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest2
```

That results in only the pods in the `kltest1` namespace being injected:

```
kubectl get po --all-namespaces | grep kltest
kltest1       authors-55d8c75c67-2sbj7                     2/2       Running   0          23m
kltest1       books-6bc5d6c547-nlhqm                       2/2       Running   0          23m
kltest1       traffic-5fccc89fbd-hgkns                     2/2       Running   0          23m
kltest1       webapp-766f6d4977-5hwrb                      2/2       Running   0          23m
kltest1       webapp-766f6d4977-nfnkb                      2/2       Running   0          23m
kltest1       webapp-766f6d4977-rvkfw                      2/2       Running   0          23m
kltest2       authors-6b757cd4fd-f4dn9                     1/1       Running   0          23m
kltest2       books-7c79648958-wm8s7                       1/1       Running   0          23m
kltest2       traffic-66b46679f9-xpr5q                     1/1       Running   0          23m
kltest2       webapp-58d887ff9d-5wxbg                      1/1       Running   0          23m
kltest2       webapp-58d887ff9d-kjwj4                      1/1       Running   0          23m
kltest2       webapp-58d887ff9d-zknjj                      1/1       Running   0          23m
```

---

Then I tear everything down:

```
curl https://run.linkerd.io/booksapp.yml | kubectl delete -f - -n kltest2
curl https://run.linkerd.io/booksapp.yml | kubectl delete -f - -n kltest1
linkerd install --proxy-auto-inject --tls optional | kubectl delete -f -
kubectl delete mutatingwebhookconfigurations/linkerd-proxy-injector-webhook-config
kubectl get cm,secret -n kltest1 -oname | grep linkerd | xargs kubectl delete -n kltest1
kubectl get cm,secret -n kltest2 -oname | grep linkerd | xargs kubectl delete -n kltest2

```

And reinstall from this branch:

```
bin/docker-build
bin/linkerd install --proxy-auto-inject --tls optional | kubectl apply -f -
```

And then apply the booksapp config in both namespaces:

```
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest1
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest2
```

And that unfortunately results in pods in both namespaces being injected:

```
kubectl get po --all-namespaces | grep kltest
kltest1       authors-5c7d5c6c7d-xk8c2                     2/2       Running   0          1m
kltest1       books-79fdf678b-qddp7                        2/2       Running   0          1m
kltest1       traffic-59949bdb85-h9k4m                     2/2       Running   0          1m
kltest1       webapp-9f7b88c7b-lm2pd                       2/2       Running   0          1m
kltest1       webapp-9f7b88c7b-nmkxd                       2/2       Running   0          1m
kltest1       webapp-9f7b88c7b-q7zlc                       2/2       Running   0          1m
kltest2       authors-5c7d5c6c7d-q8j4j                     2/2       Running   0          1m
kltest2       books-79fdf678b-qlc2h                        2/2       Running   0          1m
kltest2       traffic-59949bdb85-6mxmb                     2/2       Running   0          1m
kltest2       webapp-9f7b88c7b-j2k86                       2/2       Running   0          1m
kltest2       webapp-9f7b88c7b-prfrz                       2/2       Running   0          1m
kltest2       webapp-9f7b88c7b-r56sr                       2/2       Running   0          1m
```

---

So it seems like something related to this branch, but as I said, having stared at this code for a while, I can't figure out where the issue is coming from. I'd be fine with merging this branch and following up with a fix separately, so as not to delay merging longer than needed.

Reply:
I'm unable to reproduce on this branch. Confirming versions:

```bash
$ bin/linkerd version
Client version: git-54b6c038
Server version: git-54b6c038

$ kubectl version --short
Client Version: v1.10.7
Server Version: v1.10.11
```

```bash
$ kubectl get ns kltest1 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{"linkerd.io/inject":"enabled"},"name":"kltest1","namespace":""}}
    linkerd.io/inject: enabled
  creationTimestamp: 2019-03-04T19:39:11Z
  name: kltest1
  resourceVersion: "334201"
  selfLink: /api/v1/namespaces/kltest1
  uid: 2f1b39ea-3eb5-11e9-a352-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

$ kubectl get ns kltest2 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{},"name":"kltest2","namespace":""}}
  creationTimestamp: 2019-03-04T19:39:09Z
  name: kltest2
  resourceVersion: "334170"
  selfLink: /api/v1/namespaces/kltest2
  uid: 2de9c7bd-3eb5-11e9-a352-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

```bash
bin/linkerd install --proxy-auto-inject --tls optional | kubectl apply -f -
bin/linkerd check
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest1
curl https://run.linkerd.io/booksapp.yml | kubectl apply -f - -n kltest2
```

```bash
$ kubectl get po --all-namespaces | grep kltest
kltest1       authors-5c7d5c6c7d-7cs86                     2/2       Running   0          2m
kltest1       books-79fdf678b-p2zpw                        2/2       Running   0          2m
kltest1       traffic-59949bdb85-vvv8x                     2/2       Running   0          2m
kltest1       webapp-9f7b88c7b-hp8c9                       2/2       Running   0          2m
kltest1       webapp-9f7b88c7b-stb5q                       2/2       Running   0          2m
kltest1       webapp-9f7b88c7b-xjqjc                       2/2       Running   0          2m
kltest2       authors-6b757cd4fd-5nd5q                     1/1       Running   0          2m
kltest2       books-7c79648958-pv2hp                       1/1       Running   0          2m
kltest2       traffic-66b46679f9-9ztnj                     1/1       Running   0          2m
kltest2       webapp-58d887ff9d-2g5z7                      1/1       Running   0          2m
kltest2       webapp-58d887ff9d-cxxnp                      1/1       Running   0          2m
kltest2       webapp-58d887ff9d-pmxtd                      1/1       Running   0          2m
```

Reply:
@siggy Can you try testing against a namespace that has 0 annotations? It looks like your `kltest2` namespace still includes the `kubectl.kubernetes.io/last-applied-configuration` annotation. Likewise, I'll try testing against a namespace with at least 1 non-`linkerd.io/inject` annotation.

Reply:
Confirmed reproduction. I was using `kubectl apply` (which adds `kubectl.kubernetes.io/last-applied-configuration`) rather than `kubectl create` when creating the namespaces. When the namespace objects do not have any annotations on them, both apps get injected.

### kltest1-ns.yml

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: kltest1
  annotations:
    linkerd.io/inject: enabled
```

### kltest2-ns.yml

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: kltest2
```

## Use `kubectl create` to create namespaces

```bash
$ kubectl create -f kltest1-ns.yml
$ kubectl create -f kltest2-ns.yml

$ kubectl get ns kltest1 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    linkerd.io/inject: enabled
  creationTimestamp: 2019-03-04T19:51:50Z
  name: kltest1
  resourceVersion: "335737"
  selfLink: /api/v1/namespaces/kltest1
  uid: f3da6503-3eb6-11e9-a352-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

$ kubectl get ns kltest2 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: 2019-03-04T19:51:53Z
  name: kltest2
  resourceVersion: "335744"
  selfLink: /api/v1/namespaces/kltest2
  uid: f52f9c3c-3eb6-11e9-a352-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

## Observe both namespaces injected.

```bash
$ kubectl get po --all-namespaces | grep kltest
kltest1       authors-5c7d5c6c7d-fhqwm                     0/2       Init:0/1          0          21s
kltest1       books-79fdf678b-mhxxr                        0/2       Init:0/1          0          18s
kltest1       traffic-59949bdb85-2cnw6                     0/2       Pending           0          16s
kltest1       webapp-9f7b88c7b-fqq7x                       0/2       PodInitializing   0          21s
kltest1       webapp-9f7b88c7b-k5vsm                       0/2       Init:0/1          0          21s
kltest1       webapp-9f7b88c7b-n6kfh                       0/2       Init:0/1          0          21s
kltest2       authors-5c7d5c6c7d-l44hw                     0/2       Pending           0          8s
kltest2       books-79fdf678b-rmr7d                        0/2       Pending           0          7s
kltest2       traffic-59949bdb85-9jds6                     0/2       Pending           0          6s
kltest2       webapp-9f7b88c7b-nqxkw                       0/2       Pending           0          8s
kltest2       webapp-9f7b88c7b-qcjbg                       0/2       Pending           0          8s
kltest2       webapp-9f7b88c7b-w7vwt                       0/2       Pending           0          8s
```

Reply:
Sure enough, if I apply a config in the `kltest3` namespace, which includes a miscellaneous annotation:

```
kubectl get ns/kltest3 -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kltest: "3"
  creationTimestamp: 2019-03-04T19:52:40Z
  name: kltest3
  resourceVersion: "574251"
  selfLink: /api/v1/namespaces/kltest3
  uid: 11bb4b1e-3eb7-11e9-8d00-025000000001
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

The pods in that namespace are not inject:

```
kltest3       authors-6b757cd4fd-hx7ls                     1/1       Running   0          1m
kltest3       books-7c79648958-nwqxt                       1/1       Running   0          1m
kltest3       traffic-66b46679f9-9p6gw                     1/1       Running   0          1m
kltest3       webapp-58d887ff9d-qc2gf                      1/1       Running   0          1m
kltest3       webapp-58d887ff9d-qjtlz                      1/1       Running   0          1m
kltest3       webapp-58d887ff9d-vkvxw                      1/1       Running   0          1m
```

So it seems to only be an issue when the target namespace has 0 annotations.

Reply:
I can check to see if #2433 fixes the issue, once it has been rebased.

Reply:
Confirmed #2433 fixes the issue (also found a simpler repro via `kubectl create`:

# #2334

```bash
$ bin/linkerd install --proxy-auto-inject --tls optional | kubectl apply -f -
$ curl https://run.linkerd.io/emojivoto.yml | kubectl create -f -
$ kubectl -n emojivoto get po
NAME                        READY     STATUS    RESTARTS   AGE
emoji-7d4b89647b-9pxbg      2/2       Running   0          34s
vote-bot-846d8846fb-s9vxw   2/2       Running   0          33s
voting-5f54984cdf-nrhvs     2/2       Running   0          34s
web-6457545bcd-8vmw9        2/2       Running   0          34s
```

# #2433

```bash
$ bin/linkerd install --proxy-auto-inject --tls optional | kubectl apply -f -
$ curl https://run.linkerd.io/emojivoto.yml | kubectl create -f -
$ kubectl -n emojivoto get po
NAME                        READY     STATUS    RESTARTS   AGE
emoji-f449cd7d8-dlt8c       1/1       Running   0          7s
vote-bot-644ddf8596-62582   1/1       Running   0          6s
voting-6cfd45c845-fm9gc     1/1       Running   0          7s
web-6cddf54f66-dxbtz        1/1       Running   0          7s
```